benchmarks/prescribe_dike_injection/prescribed_dike_injection.cc

```diff
@@ -0,0 +1,571 @@
+              if (in.temperature[i] <= T_bottom_dike
+                  && this->get_timestep_number() > 0)
+                dike_injection_rate[i] = this->convert_output_to_years()
+                                         ? injection_function.value(in.position[i]) / year_in_seconds
+                                         : injection_function.value(in.position[i]);
+              else
+                dike_injection_rate[i] = 0.0;
+
+              // Dike injection effect removal
+              if (dike_injection_rate[i] > 0.0
+                  && this->get_timestep_number() > 0
+                  && in.current_cell.state() == IteratorState::valid
+                  && injection_function.value(in.current_cell->center()) > 0.0)
+                in_corrected_strainrate.strain_rate[i][0][0] -= dike_injection_rate[i];
+            }
+        }
+      
+      // Fill variable out with the results form the base material model
+      // using the corrected model inputs.
+      base_model->evaluate(in_corrected_strainrate, out);
+
+      // Below we start to track the motion of the dike injection material.
+      AssertThrow(this->introspection().compositional_name_exists("injection_phase"),
+                  ExcMessage("Material model 'prescribed dike injection' only works if "
+                             "there is a compositional field called 'injection_phase'. "));
+
+      // Index for injection phase
+      unsigned int injection_phase_index = this->introspection().compositional_index_for_name("injection_phase");
+
+      // Indices for all chemical compositional fields, and not e.g., plastic strain.
+      const std::vector<unsigned int> chemical_composition_indices = this->introspection().chemical_composition_field_indices();
+
+      const auto &component_indices = this->introspection().component_indices.compositional_fields;
+
+      // Positions of quadrature points at the current cell
+      std::vector<Point<dim>> quadrature_positions;
+
+      // The injection material will replace part of the original material 
+      // based on the injection rate and diking duration, i.e., fraction 
+      // of injected material to original material for the existence 
+      // duration of a dike.
+      double dike_injection_fraction = 0.0;
+
+      for (unsigned int i=0; i < in.n_evaluation_points(); ++i)
+        {
+          // Activate the dike injection by adding the additional RHS 
+          // terms of injection to Stokes equations.
+          if (prescribed_dilation != nullptr)
+            prescribed_dilation->dilation[i] = dike_injection_rate[i];
+
+          // User-defined or timestep-dependent injection fraction.
+          if (this->simulator_is_past_initialization())
+            dike_injection_fraction = dike_injection_rate[i] * this->get_timestep();
+
+          if (dike_material_injection_fraction != 0.0)
+            dike_injection_fraction = dike_material_injection_fraction;
```

Comment:
This requires some more comments. This can set the dike_injection_fraction to the user-set fraction, even if the dike_injection_rate is zero, but because below we only fill the reaction_rates for a dike_injection_rate >0, this is okay.

Is the dike injection fraction consistent the with injection_rate if the fraction is user-set?